lib/util/source-code.js

```diff
@@ -223,14 +224,110 @@ class SourceCode extends TokenStore {
     }
 
     /**
+     * Gets all comments before the given node or token.
+     * @param {ASTNode|token} nodeOrToken The AST node or token to get the comments for.
+     * @returns {Object} The list of comments indexed by their position.
+     * @public
+     */
+    getCommentsBefore(nodeOrToken) {
```

Comment:
It might be more performant to add this to `TokenStore` rather than `SourceCode`, because `TokenStore` can reuse a cursor while iterating backwards rather than creating a new cursor every time `getTokenBefore` is called.